api/server_test.go

```diff
@@ -136,22 +45,239 @@ func TestServerRoutes(t *testing.T) {
+		},
+		{"POST", "/add", "testdata/addResponseFromFile.json", 200,
+			map[string]string{
+				"file":      "testdata/cities/data.csv",
+				"structure": "testdata/cities/structure.json",
+				"metadata":  "testdata/cities/meta.json",
+			},
+			map[string]string{
+				"peername": "peer",
+				"name":     "cities",
+			},
+		},
+		{"POST", "/save", "testdata/saveResponse.json", 200,
+			map[string]string{
+				"file": "testdata/cities/data_update.csv",
+			},
+			map[string]string{
+				"peername": "peer",
+				"name":     "cities",
+				"title":    "added row to include Seoul, Korea",
+				"message":  "want to expand this list to include more cities",
+			},
+		},
+		{"POST", "/save", "testdata/saveResponseMeta.json", 200,
+			map[string]string{
+				"metadata": "testdata/cities/meta_update.json",
+			},
+			map[string]string{
+				"peername": "peer",
+				"name":     "cities",
+				"title":    "Adding more specific metadata",
+				"message":  "added title and keywords",
+			},
+		},
+		{"POST", "/profile/photo", "testdata/photoResponse.json", 200,
+			map[string]string{
+				"file": "testdata/rico_400x400.jpg",
+			},
+			map[string]string{
+				"peername": "peer",
+			},
+		},
+		{"POST", "/profile/poster", "testdata/posterResponse.json", 200,
+			map[string]string{
+				"file": "testdata/rico_poster_1500x500.jpg",
+			},
+			map[string]string{
+				"peername": "peer",
+			},
+		},
+	}
+
+	for i, c := range cases {
+
+		expectBody, err := ioutil.ReadFile(c.expectBodyPath)
+		if err != nil {
+			t.Errorf("case add dataset from file, error reading expected response from file: %s", err)
+		}
+
+		req, err := MakeMimeMultipartReq(c.method, c.endpoint, server.URL+c.endpoint, c.filePaths, c.params)
+		if err != nil {
+			t.Errorf("testMimeMultipart case %d, %s - %s:\nerror making mime/multipart request: %s", i, c.method, c.endpoint, err)
+			continue
+		}
+
+		res, err := client.Do(req)
+		if err != nil {
+			t.Errorf("testMimeMultipart case %d, %s - %s:\nerror performing request: %s", i, c.method, c.endpoint, err)
+			continue
+		}
+
+		gotBody, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			t.Errorf("testMimeMultipart case %d, %s - %s:\nerror reading response body request: %s", i, c.method, c.endpoint, err)
+			continue
+		}
+
+		if string(gotBody) != string(expectBody) {
+			t.Errorf("testMimeMultipart case %d, %s - %s:\nresponse body mismatch. expected: %s, got %s", i, c.method, c.endpoint, string(expectBody), string(gotBody))
+			continue
+		}
+
+		if res.StatusCode != 200 {
+			t.Errorf("testMimeMultipart case %d, %s - %s:\nstatus code mismatch. expected: %d, got: %d", i, c.method, c.endpoint, 200, res.StatusCode)
+			continue
+		}
+	}
+}
+
+// MakeMimeMultipartBody takes a map of filepaths and params and add thems to a writer
+// adds that writer to *bytes.Buffer that can then be used as the body for a request
+func MakeMimeMultipartReq(method, endpoint, url string, filePaths, params map[string]string) (*http.Request, error) {
```

Comment:
consider using `NewFilesRequest`. Also, I would reword the comment:
```golang
// NewFilesRequest creates a mime/multipart http.Request with files specified by a map of param : filepath,
// and form values specified by a map, params
```
When writing comments, try to think about the consumer of the function instead of it's internals.